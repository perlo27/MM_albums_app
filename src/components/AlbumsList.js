import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { FlatList, View, Text, StyleSheet } from 'react-native';
import { ListItem, Header, Title, Spinner, Left, Right, Icon } from 'native-base';
import { Transition } from 'react-navigation-fluid-transitions';

import { withDataLoader } from '../decorators';
import { navigationPaths } from '../navigator';
import { album } from '../propTypes';
import { headerHeight } from '../config';

@withDataLoader
export class AlbumsList extends PureComponent {
  static propTypes = {
    albums: PropTypes.arrayOf(album),
  };

  static defaultProps = {
    albums: [],
  };

  onPress = (id, title) => () =>
    this.props.navigation.navigate(navigationPaths.album, { id, title });

  keyExtractor = item => `${item.id}`;

  renderAlbum = ({ item: { title, id } }) => (
    <ListItem onPress={this.onPress(id, title)}>
      <Left>
        <Text>{title}</Text>
      </Left>
      <Right>
        <Icon name="arrow-forward" />
      </Right>
    </ListItem>
  );

  render() {
    const { albums } = this.props;
    return (
      <Transition appear="horizontal">
        <View>
          <Header style={styles.header}>
            <Title>Albums</Title>
          </Header>
          <FlatList
            data={albums}
            keyExtractor={this.keyExtractor}
            renderItem={this.renderAlbum}
            ListEmptyComponent={<Spinner color="green" />}
          />
        </View>
      </Transition>
    );
  }
}

const styles = StyleSheet.create({
  header: {
    backgroundColor: 'white',
    height: headerHeight,
    paddingTop: 0,
    alignItems: 'center',
  },
});
