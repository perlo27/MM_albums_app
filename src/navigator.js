import { FluidNavigator } from 'react-navigation-fluid-transitions';
import React from 'react';
import { View, Text } from 'react-native';

import { AlbumsList, Album } from './components';

export const paths = {
  albumList: 'albumList',
  album: 'album',
  photo: 'photo',
};

export default FluidNavigator({
  [paths.albumList]: {
    screen: AlbumsList,
  },
  [paths.album]: Album,
  [paths.photo]: {
    screen: () => (
      <View>
        <Text>photo</Text>
      </View>
    ),
  },
});
