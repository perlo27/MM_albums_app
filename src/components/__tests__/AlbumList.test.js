import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { AlbumsList } from '../AlbumsList';
import { navigation, store } from './moks';
import { navigationPaths } from '../../navigator';

jest.mock('../Header', () => ({
  Header: 'Header',
}));

describe('AlbumsList', () => {
  it('should render correctly', () => {
    const tree = renderer.create(<AlbumsList navigation={navigation} store={store} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should navigate to album page on listItem press', () => {
    const wrapper = shallow(<AlbumsList navigation={navigation} store={store} />);

    const testItem = { item: { title: 'test', id: 1 } };
    const a = wrapper
      .dive()
      .find('FlatList')
      .first()
      .prop('renderItem')(testItem);

    a.props.onPress();
    expect(navigation.navigate).toHaveBeenCalledWith(navigationPaths.album, testItem.item);
  });
});
