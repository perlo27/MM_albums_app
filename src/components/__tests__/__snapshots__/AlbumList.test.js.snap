// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlbumsList should render correctly 1`] = `
<Transition
  appear="horizontal"
>
  <View>
    <Header
      style={
        Array [
          Object {
            "alignItems": "center",
            "height": 50,
            "paddingTop": 0,
          },
          Object {
            "backgroundColor": "white",
          },
        ]
      }
    >
      <Title>
        Albums
      </Title>
    </Header>
    <RCTScrollView
      ListEmptyComponent={
        <Spinner
          color="green"
        />
      }
      data={
        Array [
          Object {
            "id": 1,
            "title": "Quidem molestiae enim",
            "userId": 1,
          },
          Object {
            "id": 2,
            "title": "Quidem molestiae enim 1",
            "userId": 2,
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <ListItem
            onPress={[Function]}
          >
            <Left>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
              >
                Quidem molestiae enim
              </Text>
            </Left>
            <Right>
              <Icon
                name="arrow-forward"
              />
            </Right>
          </ListItem>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <ListItem
            onPress={[Function]}
          >
            <Left>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
              >
                Quidem molestiae enim 1
              </Text>
            </Left>
            <Right>
              <Icon
                name="arrow-forward"
              />
            </Right>
          </ListItem>
        </View>
      </View>
    </RCTScrollView>
  </View>
</Transition>
`;
